zipline/pipeline/loaders/utils.py

```diff
@@ -260,3 +274,53 @@ def check_data_query_args(data_query_time, data_query_tz):
                 data_query_tz,
             ),
         )
+
+
+def zip_with_floats(dates, flts):
+        return pd.Series(flts, index=dates, dtype='float')
+
+
+def num_days_between(dates, start_date, end_date):
+    return num_days_in_range(dates, start_date, end_date)
+
+
+def zip_with_dates(index_dates, dts):
+    return pd.Series(pd.to_datetime(dts), index=index_dates)
+
+
+def get_values_for_date_ranges(zip_date_index_with_vals,
+                               vals_for_date_intervals,
+                               date_intervals,
+                               date_index):
+    """
+    Returns a Series of values indexed by date based on values for the given
+    date intervals.
+
+    Parameters
+    ----------
+    zip_date_index_with_vals : callable
+        A function that takes in a list of dates and a list of values and
+        returns a pd.Series with the values indexed by the dates.
+    vals_for_date_intervals : list
+        A list of values for each date interval in `date_intervals`.
+    date_intervals : list
+        A list of pairs of dates, where each pair represents a date interval
+        that corresponds to the value at the same index in
+        `vals_for_date_intervals`.
+    date_index : DatetimeIndex
+        The DatetimeIndex containing all dates for which values were requested.
+
+    Returns
+    -------
+    date_index_with_vals : pd.Series
+        A Series indexed by the given DatetimeIndex and with values assigned
+        to dates based on the given date intervals.
+    """
+    # Fill in given values for given date ranges.
+    return zip_date_index_with_vals(
+        date_index,
+        np.repeat(vals_for_date_intervals,
```

Comment:
Big :+1: on the usage of `repeat` here.

If you look at the implementation of `num_days_in_range` (which is equivalent to `num_days_between`, see comment above), you'll see that all it's doing is a sorted lookup into the DatetimeIndex with the `start, end` pair you're providing, and then subtracting the difference.  So the this whole expression is essentially doing something like:
```
In [31]: dates = pd.date_range('1990', '1990-01-15', freq='2D')
In [32]: dates
Out[32]:
DatetimeIndex(['1990-01-01', '1990-01-03', '1990-01-05', '1990-01-07',
               '1990-01-09', '1990-01-11', '1990-01-13', '1990-01-15'],
              dtype='datetime64[ns]', freq='2D', tz=None)

In [33]: values = array([1, 2, 3])

In [34]: bounds = [(Timestamp('1990-01-02'), Timestamp('1990-01-06')),
                  (Timestamp('1990-01-03'), Timestamp('1990-01-08')),
                  (Timestamp('1990-01-11'), Timestamp('1990-01-14'))]

In [35]: deltas = [data.searchsorted(end) - data.searchsorted(start) for start, end in bounds]

In [36]: deltas
Out[36]: [2, 3, 2]

In [37]: np.repeat(values, deltas)
Out[37]: array([1, 1, 2, 2, 2, 3, 3])
```

The only thing here that's still happening iteratively in Python is the slicing to determine the number of repeats for each value you have.  You could make that vectorized as well by passing all the starts at once to `searchsorted`, then passing all the ends, and then doing a vectorized subtraction.  With the input types you have right now, this would look something like:

```
dates.values.searchsorted([b[1].asm8 for b in bounds]) - dates.values.searchsorted([b[0].asm8 for b in bounds])
```
which probably isn't much faster (in fact it would probably be slower until the inputs got very large), because we have to coerce the datetimeindex to an array and coerce all the Timestamps to arrays of `datetime64[ns]` to do the broadcasted lookups.  If this function turns out to be a bottleneck, we'd probably want to change the input formats for the bounds from a list of pairs to two parallel arrays of `(start_date, end_date)`.  Unless it's very straightforward to do so, that change probably isn't worth making without more profiling.